libfreerdp/core/gateway/rdg.c

```diff
@@ -323,34 +359,346 @@ static BOOL rdg_write_packet(rdpRdg* rdg, wStream* sPacket)
+		return 0;
+	}
+
+	status =
+	    BIO_read(bio, pBuffer,
+	             (encodingContext->payloadLength < size ? encodingContext->payloadLength : size));
+	if (status <= 0)
+		return status;
+
+	encodingContext->payloadLength -= status;
+
+	if (encodingContext->payloadLength == 0)
+		encodingContext->state = WebsocketStateOpcodeAndFin;
+
+	return status;
+}
+
+static int rdg_websocket_read_discard(BIO* bio, rdg_http_websocket_context* encodingContext)
+{
+	char _dummy[256];
+	int status;
+
+	if (encodingContext->payloadLength == 0)
+	{
+		encodingContext->state = WebsocketStateOpcodeAndFin;
+		return 0;
+	}
+
+	status = BIO_read(bio, _dummy, sizeof(_dummy));
+	if (status <= 0)
+		return status;
+
+	encodingContext->payloadLength -= status;
+
+	if (encodingContext->payloadLength == 0)
+		encodingContext->state = WebsocketStateOpcodeAndFin;
+
+	return status;
+}
+
+static int rdg_websocket_reply_close(BIO* bio)
+{
+	/* write back close */
+	wStream* closeFrame;
+	uint32_t maskingKey;
+	int status;
+
+	closeFrame = Stream_New(NULL, 6);
+	Stream_Write_UINT8(closeFrame, WEBSOCKET_FIN_BIT | WebsocketPongOpcode);
+	Stream_Write_UINT8(closeFrame, 0 | WEBSOCKET_MASK_BIT); /* no payload */
+	winpr_RAND((BYTE*)&maskingKey, 4);
+	Stream_Write_UINT32(closeFrame, maskingKey); /* dummy masking key. */
+	Stream_SealLength(closeFrame);
+
+	status = BIO_write(bio, Stream_Buffer(closeFrame), Stream_Length(closeFrame));
+	/* server MUST close socket now. The server is not allowed anymore to
+	 * send frames but if he does, nothing bad would happen */
+	return status;
+}
+
+static int rdg_websocket_reply_pong(BIO* bio)
+{
+	wStream* pong;
+	uint32_t maskingKey;
+	int status;
+	/* write pong back */
+	pong = Stream_New(NULL, 6);
+	Stream_Write_UINT8(pong, WEBSOCKET_FIN_BIT | WebsocketPongOpcode);
+	Stream_Write_UINT8(pong, 0 | WEBSOCKET_MASK_BIT); /* no payload */
+	winpr_RAND((BYTE*)&maskingKey, 4);
+	Stream_Write_UINT32(pong, maskingKey); /* dummy masking key. */
+	Stream_SealLength(pong);
+
+	status = BIO_write(bio, Stream_Buffer(pong), Stream_Length(pong));
+	Stream_Free(pong, TRUE);
+
+	return status;
+}
+
+static int rdg_websocket_handle_payload(BIO* bio, BYTE* pBuffer, size_t size,
+                                        rdg_http_websocket_context* encodingContext)
+{
+	int status;
+	BYTE effectiveOpcode = ((encodingContext->opcode & 0xf) == WebsocketContinuationOpcode
+	                            ? encodingContext->fragmentOriginalOpcode & 0xf
+	                            : encodingContext->opcode & 0xf);
+
+	switch (effectiveOpcode)
+	{
+		case WebsocketBinaryOpcode:
+		{
+			status = rdg_websocket_read_data(bio, pBuffer, size, encodingContext);
+			if (status < 0)
+				return status;
+
+			return status;
+		}
+		break;
+		case WebsocketPingOpcode:
+		{
+			status = rdg_websocket_read_discard(bio, encodingContext);
+			if (status < 0)
+				return status;
+
+			if (encodingContext->payloadLength == 0)
+				rdg_websocket_reply_pong(bio);
```

Comment:
According to the rfc section  5.5.3 (https://tools.ietf.org/html/rfc6455#section-5.5.3) the pong answer must have identical "Application data" if there is "Application data" included in the ping frame.

Reply:
I will look into this. Do you have a way to test this? Afaik MS does not send pings, so introducing a buffer for this requires someone to test it or at least confirm that it will work (probably more that 4 eyes looking on this).
https://tools.ietf.org/html/rfc6455#section-5.5.1 (close) does not mandate relaying back application data.

Reply:
Hi, the rfc just states that it would be typically to echo the status code of the close:
`(When
   sending a Close frame in response, the endpoint typically echos the
   status code it received.) `

Reply:
I did a Test with a quick and got it to echo the ping/pong back (by patching a simple websocket server/gateway to send a ping immediately after the websocket upgrade). Hope you like the way I implemented the buffer handling.
I even implemented relaying back the status code of the close frame and enforced that no data is sent after that.

Reply:
@akarl10 very nice job!